src/main/java/io/r2dbc/postgresql/client/SSLSessionHandlerAdapter.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2017-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.r2dbc.postgresql.client;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufAllocator;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelInboundHandlerAdapter;
+import io.netty.handler.ssl.SslHandler;
+import io.r2dbc.postgresql.message.frontend.SSLRequest;
+import reactor.core.publisher.Mono;
+import reactor.netty.tcp.SslProvider;
+
+import java.util.concurrent.CompletableFuture;
+
+public class SSLSessionHandlerAdapter extends ChannelInboundHandlerAdapter {
+
+    private final ByteBufAllocator alloc;
+
+    private final SslHandler sslHandler;
+
+    private final CompletableFuture<Void> handshakeFuture;
+
+    public SSLSessionHandlerAdapter(SslProvider sslProvider, ByteBufAllocator alloc) {
+        this.alloc = alloc;
+        this.sslHandler = new SslHandler(sslProvider.getSslContext().newEngine(alloc));
+        this.handshakeFuture = new CompletableFuture<>();
+        this.sslHandler.handshakeFuture().addListener(future -> {
+            if (future.isSuccess()) {
+                this.handshakeFuture.complete(null);
+            } else {
+                this.handshakeFuture.completeExceptionally(future.cause());
+            }
+        });
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) {
+        ByteBuf buf = (ByteBuf) msg;
+        char response = (char) buf.readByte();
+        switch (response) {
+            case 'S':
+                ctx.channel().pipeline()
+                    .addFirst(this.sslHandler)
+                    .remove(this);
+                ctx.fireChannelRead(msg);
+                break;
+            case 'N':
+                throw new IllegalStateException("Server doesn't support SSL connections");
```

Comment:
Do we need throwing exceptions? Why don't we complete `handshakeFuture` exceptionally?